src/common/datas/holder.ts

```diff
@@ -0,0 +1,174 @@
+
+  /**
+   * Sets a boolean.
+   * @param key Key.
+   * @param type 'boolean'
+   * @param value New value.
+   */
+  abstract set(key: string, type: 'boolean', value: boolean): void;
+
+  /**
+   * Sets an integer or a number. In this case, an integer is a 32-bit signed
+   * integer. A number is standard JavaScript 64-bit floating point number.
+   * @param key Key.
+   * @param type 'integer' or 'boolean'.
+   * @param value New value.
+   */
+  abstract set(key: string, type: 'integer' | 'number', value: number): void;
+
+  /**
+   * Sets a string.
+   * @param key Key.
+   * @param type 'string'.
+   * @param value New value.
+   */
+  abstract set(key: string, type: 'string', value: string): void;
+}
+
+export function dataHolder(storage: PersistentDataHolder): DataHolder {
+  // TODO other storage types
+  return new BukkitHolder(storage.getPersistentDataContainer());
+}
+
+function fromTypedJson<T>(
+  type: new () => T,
+  json: string | null,
+): T | undefined {
+  if (json == null) {
+    return undefined;
+  }
+  const data = JSON.parse(json);
+  if (data.t != type.name) {
+    throw new Error(`type in JSON is ${data.t}, ${type.name} expected`);
+  }
+  return data.d;
+}
+
+function toTypedJson<T>(type: new () => T, data: T): string {
+  return JSON.stringify({ t: type.name, d: data });
+}
+
+function intToBoolean(value: Integer | null): boolean | null {
+  return value != null ? ((value as unknown) as number) != 0 : null;
+}
+
+function namespacedKey(key: string): NamespacedKey {
+  return new NamespacedKey(__plugin, key);
+}
+
+/**
+ * Holds data in Spigot's PersistentDataContainer.
+ */
+class BukkitHolder extends DataHolder {
+  container: PersistentDataContainer;
+
+  constructor(container: PersistentDataContainer) {
+    super();
+    this.container = container;
+  }
+
+  get(key: string, type: any): any {
+    const containerKey = namespacedKey(key);
+    switch (type) {
+      case 'boolean':
+        return intToBoolean(
+          this.container.get(containerKey, PersistentDataType.INTEGER),
+        );
+      case 'integer':
+        return this.container.get(containerKey, PersistentDataType.INTEGER);
+      case 'number':
+        return this.container.get(containerKey, PersistentDataType.DOUBLE);
+      case 'string':
+        return this.container.get(containerKey, PersistentDataType.STRING);
+      default:
+        return fromTypedJson(
+          type,
+          this.container.get(containerKey, PersistentDataType.STRING),
+        );
+    }
+  }
+
+  set(key: string, type: any, value: any): void {
+    const containerKey = namespacedKey(key);
+    switch (type) {
+      case 'boolean':
+        this.container.set(
+          containerKey,
+          PersistentDataType.INTEGER,
+          ((value ? 1 : 0) as unknown) as Integer,
+        );
+        break;
+      case 'integer':
+        this.container.set(containerKey, PersistentDataType.INTEGER, value);
```

Comment:
This probably throws an error if the `value` is not actually an integer? If so, the "Java-exception" should probably be handled and the code should throw a "JS-exception", i.e. an `Error`-object. Or should it just round it an pretend nothing happened?

Reply:
I've added check to ensure it is actually an integer.